MAINT use _valid_parameter in TfidfTransformer

brandonthchen opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs 
towards  https://github.com/scikit-learn/scikit-learn/issues/23462
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#### What does this implement/fix? Explain your changes.

Use _validate_params in TfidfTransformer 
fit calls _validate_parameters as the first step
remove TfidfTransformer in  PARAM_VALIDATION_ESTIMATORS_TO_IGNORE in  scikit-learn/sklearn/tests/test_common.py

#### Any other comments?

This is my first contribution. Let me know if there is anything not so right. 


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

brandonthchen committed:
TfidfTransformer Use _validate_params in TfidfTransformer

brandonthchen commented:
There is an errors in the testing .   
1. It looks like that the test file put None into norm parameter in TfidfVectorizer and ,then, None flows into TfidfTransformer .  I can add None  into _validate_params but should I do that  ?  I think Norm should be either l1 or l2. 

>   ../1/s/sklearn/feature_extraction/tests/test_text.py:1290: 
       def test_tfidfvectorizer_binary():
              # Non-regression test: TfidfVectorizer used to ignore its "binary" param.
              v = TfidfVectorizer(binary=True, use_idf=False, norm=None)
              assert v.binary

jeremiedbb commented on sklearn/feature_extraction/text.py:
```diff
@@ -1594,6 +1594,15 @@ class TfidfTransformer(_OneToOneFeatureMixin, TransformerMixin, BaseEstimator):
     """
 
+    _parameter_constraints = {
+        "norm": [
+            StrOptions({"l1", "l2"})
+        ],
```
This fits in one line
```suggestion
        "norm": [StrOptions({"l1", "l2"})],
```

jeremiedbb reviewed (commented):
Thanks for the PR @brandonthchen.

jeremiedbb commented:
> There is an errors in the testing. I have an question about that error.

> It looks like that the test file put None into norm parameter in TfidfVectorizer and ,then, None flows into TfidfTransformer . I can add None into _validate_params but should I do that ? I think Norm should be either l1 or l2.

Looking at the code of `transform`, it looks like `norm=None` is a valid use case, meaning that we don't normalize the vectors at the end. You can add `None` into the valid options. Can you also update the docstring to make it accurate ?

jeremiedbb commented:
Also, looks like you have linting issues. You can fix it by installing and running `black`

brandonthchen committed:
TfidfTransformer Use _validate_params in TfidfTransformer

brandonthchen committed:
TfidfTransformer Use _validate_params in TfidfTransformer

brandonthchen commented:
Thanks [jeremiedbb](https://github.com/jeremiedbb) for the review.  I have fixed the linting issue and added None into parameter validation.   

There are additional errors for this commit. I guess it is because I didn't properly add the docstring .  This is my first time to modify docstring .  Is there any guide/documentation to help me updating the docstring ?

jeremiedbb committed:
Merge branch 'main' into TfidfTransformer_Use_validate_params_in_TfidfTransformer

jeremiedbb commented on sklearn/feature_extraction/text.py:
```diff
@@ -1519,6 +1519,7 @@ class TfidfTransformer(_OneToOneFeatureMixin, TransformerMixin, BaseEstimator):
           been applied.
         - 'l1': Sum of absolute values of vector elements is 1.
           See :func:`preprocessing.normalize`.
+        - None: The program does not normalize the vectors at the end.
```
None should also be listed in the parameter description on line 1514:
``norm: {'l1', 'l2'} or None, default='l2'``

Also I would not put None in this bullet list, since the list describes `Each output row will have unit norm, either:`. I would just put that after the list, and maybe rephrase a bit: something like `If None, output rows are not normalized.`

jeremiedbb commented on sklearn/feature_extraction/text.py:
```diff
@@ -32,7 +32,7 @@
 from ..utils import _IS_32BIT
 from ..exceptions import NotFittedError
-
+from ..utils._param_validation import StrOptions
```
```suggestion
from ..utils._param_validation import StrOptions

```

jeremiedbb reviewed (commented):
The failure in the CI is unrelated to this PR and will be solved when https://github.com/scikit-learn/scikit-learn/pull/23558 is merged

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/brandonthchen/23483

jeremiedbb committed:
include recent changes

jeremiedbb reviewed (approved):
LGTM

glemaitre reviewed (approved):
LGTM
